python/ray/data/tests/test_operators.py

```diff
@@ -0,0 +1,199 @@
+    input_op = InputDataBuffer(_make_ref_bundles([[i] for i in range(100)]))
+    op = MapOperator(_mul2_transform, input_op=input_op, name="TestMapper")
+
+    # Feed data and block on exec.
+    while input_op.has_next():
+        op.add_input(input_op.get_next(), 0)
+    op.inputs_done(0)
+    for work in op.get_work_refs():
+        ray.get(work)
+        op.notify_work_completed(work)
+
+    # Check we return transformed bundles in order.
+    assert _take_outputs(op) == [[i * 2] for i in range(100)]
+
+    # Check dataset stats.
+    stats = op.get_stats()
+    assert "TestMapper" in stats, stats
+    assert len(stats["TestMapper"]) == 100, stats
+
+    # Check memory stats.
+    metrics = op.get_metrics()
+    assert metrics["obj_store_mem_alloc"] == pytest.approx(8800, 0.5), metrics
+    assert metrics["obj_store_mem_peak"] == pytest.approx(8800, 0.5), metrics
+    assert metrics["obj_store_mem_freed"] == pytest.approx(6400, 0.5), metrics
+
+
+def test_map_operator_streamed(ray_start_regular_shared):
+    # Create with inputs.
+    input_op = InputDataBuffer(_make_ref_bundles([[i] for i in range(100)]))
+    op = MapOperator(_mul2_transform, input_op=input_op, name="TestMapper")
+
+    # Feed data and implement streaming exec.
+    output = []
+    while input_op.has_next():
+        op.add_input(input_op.get_next(), 0)
+        for work in op.get_work_refs():
+            ray.get(work)
+            op.notify_work_completed(work)
+        assert op.has_next()
+        while op.has_next():
+            ref = op.get_next()
+            assert ref.owns_blocks, ref
+            _get_blocks(ref, output)
+
+    # Check equivalent to bulk execution in order.
+    assert output == [[i * 2] for i in range(100)]
+    metrics = op.get_metrics()
+    assert metrics["obj_store_mem_alloc"] == pytest.approx(8800, 0.5), metrics
+    assert metrics["obj_store_mem_peak"] == pytest.approx(88, 0.5), metrics
+    assert metrics["obj_store_mem_freed"] == pytest.approx(6400, 0.5), metrics
+
+
+def test_map_operator_min_rows_per_batch(ray_start_regular_shared):
+    # Simple sanity check of batching behavior.
+    def _check_batch(block_iter: Iterable[Block]) -> Iterable[Block]:
+        block_iter = list(block_iter)
+        assert len(block_iter) == 5, block_iter
+        for block in block_iter:
+            yield block
+
+    # Create with inputs.
+    input_op = InputDataBuffer(_make_ref_bundles([[i] for i in range(10)]))
+    op = MapOperator(
+        _check_batch,
+        input_op=input_op,
+        name="TestMapper",
+        min_rows_per_batch=5,
+    )
+
+    # Feed data and block on exec.
+    while input_op.has_next():
+        op.add_input(input_op.get_next(), 0)
+    op.inputs_done(0)
+    for work in op.get_work_refs():
+        ray.get(work)
+        op.notify_work_completed(work)
+
+    # Check we return transformed bundles in order.
+    assert _take_outputs(op) == [[i] for i in range(10)]
+
+
+def test_map_operator_ray_args(shutdown_only):
```

Comment:
Debating whether it's worth it to mock out the Ray API here to speed up these tests a bit. Maybe it's not that important since the bulk of the testing will be for StreamingExecutor, which we can write separate mocks for.